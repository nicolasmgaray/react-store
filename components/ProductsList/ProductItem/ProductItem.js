import React from "react";
import {
  Container,
  ProductName,
  RedeemScreen,
  ReedemButton,
  ProductImage,
  Category,
  MissingPoints,
  Points,
  BuyLogo,
  Break,
  Coin
} from "../../../styles/Products";
import Ellipse from "../../UI/Ellipse";

const Product = ({
  name,
  cost,
  category,
  img,
  redeem,
  isRedeeming,
  points,
  isHovered,
  setHover
}) => {
  return (
    <Container
      onMouseEnter={() => {
        setHover(true);
      }}
      onMouseDown={() => {
        setHover(true);
      }}
      onMouseLeave={() => {
        setHover(false);
      }}
    >
      {points >= cost ? (
        <BuyLogo
          src={
            isHovered || isRedeeming
              ? "/icons/buy-white.svg"
              : "/icons/buy-blue.svg"
          }
        />
      ) : (
        <MissingPoints>
          {`You need ${cost - points} `} <Coin />
        </MissingPoints>
      )}
      {(points >= cost && isHovered) || isRedeeming ? (
        <RedeemScreen>
          <Points>
            {cost}
            <Coin />
          </Points>
          <ReedemButton onClick={redeem}>
            {isRedeeming ? <Ellipse /> : `Redeem Now`}
          </ReedemButton>
        </RedeemScreen>
      ) : null}

      <ProductImage src={img.url} alt={name}></ProductImage>
      <Break />
      <Category>{category}</Category>
      <ProductName>{name}</ProductName>
    </Container>
  );
};

export default Product;
