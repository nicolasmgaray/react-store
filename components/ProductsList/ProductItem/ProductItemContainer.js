import React, { useState } from "react";
import Product from "./ProductItem";
import { useDispatch } from "react-redux";
import { actions } from "../../../redux/ducks/User";
import Swal from "sweetalert2";

const ProductContainer = props => {
  const dispatch = useDispatch();
  const [isRedeeming, setRedeeming] = useState(false);
  const { redeem } = actions;

  const handleRedeem = async () => {
    try {
      setRedeeming(true);
      const result = await dispatch(redeem(props._id, props.cost));
      if (result.error) throw result.error;
      Swal.fire({
        icon: "success",
        title: result.message
      });
    } catch (err) {
      Swal.fire({
        icon: "error",
        title: err
      });
    } finally {
      setRedeeming(false);
    }
  };

  return (
    <Product
      redeem={handleRedeem}
      isRedeeming={isRedeeming}
      {...props}
    ></Product>
  );
};

export default ProductContainer;
