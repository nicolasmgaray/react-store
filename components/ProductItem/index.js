import ProductItemContainer from "./ProductItemContainer";
export default ProductItemContainer;