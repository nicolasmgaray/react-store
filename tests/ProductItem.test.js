import React from "react";
import { render } from "enzyme";
import ProductItem from "../components/ProductsList/ProductItem/ProductItem.js";

describe("ProductItem", () => {
  const props = {
    name: "iPhone 8",
    cost: 800,
    category: "Phones",
    img: {
      url: "https://aerolab-challenge.now.sh/images/iPhone8-x1.png",
      hdUrl: "https://aerolab-challenge.now.sh/images/iPhone8-x2.png"
    },
    redeem: () => {},
    points: 5000,
    isHovered: false,
    isRedeeming: false,
    setHover: () => {}
  };

  it("renders without crashing", () => {
    render(<ProductItem {...props} />);
  });
  it("matches snapshot", () => {
    const component = render(<ProductItem {...props} />);
    expect(component).toMatchSnapshot();
  });
});
