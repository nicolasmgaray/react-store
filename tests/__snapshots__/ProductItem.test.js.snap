// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductItem matches snapshot 1`] = `
initialize {
  "0": Object {
    "attribs": Object {
      "class": "Products__Container-sc-189ksbx-11 bfkBuI",
    },
    "children": Array [
      Object {
        "attribs": Object {
          "class": "Products__BuyLogo-sc-189ksbx-2 bvVWja",
          "src": "/icons/buy-blue.svg",
        },
        "children": Array [],
        "name": "img",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "alt": "iPhone 8",
            "class": "Products__ProductImage-sc-189ksbx-7 fsrlVZ",
            "src": "https://aerolab-challenge.now.sh/images/iPhone8-x1.png",
          },
          "children": Array [],
          "name": "img",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {
              "class": "Products__Break-sc-189ksbx-8 hPsqUm",
            },
            "children": Array [],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "Products__Category-sc-189ksbx-9 jLhkHD",
              },
              "children": Array [
                Object {
                  "data": "Phones",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": Object {
                "attribs": Object {
                  "class": "Products__ProductName-sc-189ksbx-10 fznLgY",
                },
                "children": Array [
                  Object {
                    "data": "iPhone 8",
                    "next": null,
                    "parent": [Circular],
                    "prev": null,
                    "type": "text",
                  },
                ],
                "name": "div",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": [Circular],
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "alt": undefined,
            "class": undefined,
            "src": undefined,
          },
          "x-attribsPrefix": Object {
            "alt": undefined,
            "class": undefined,
            "src": undefined,
          },
        },
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
          "src": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
          "src": undefined,
        },
      },
      Object {
        "attribs": Object {
          "alt": "iPhone 8",
          "class": "Products__ProductImage-sc-189ksbx-7 fsrlVZ",
          "src": "https://aerolab-challenge.now.sh/images/iPhone8-x1.png",
        },
        "children": Array [],
        "name": "img",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "Products__Break-sc-189ksbx-8 hPsqUm",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {
              "class": "Products__Category-sc-189ksbx-9 jLhkHD",
            },
            "children": Array [
              Object {
                "data": "Phones",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "Products__ProductName-sc-189ksbx-10 fznLgY",
              },
              "children": Array [
                Object {
                  "data": "iPhone 8",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "text",
                },
              ],
              "name": "div",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "Products__BuyLogo-sc-189ksbx-2 bvVWja",
            "src": "/icons/buy-blue.svg",
          },
          "children": Array [],
          "name": "img",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": null,
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
            "src": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
            "src": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "alt": undefined,
          "class": undefined,
          "src": undefined,
        },
        "x-attribsPrefix": Object {
          "alt": undefined,
          "class": undefined,
          "src": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "Products__Break-sc-189ksbx-8 hPsqUm",
        },
        "children": Array [],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "Products__Category-sc-189ksbx-9 jLhkHD",
          },
          "children": Array [
            Object {
              "data": "Phones",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": Object {
            "attribs": Object {
              "class": "Products__ProductName-sc-189ksbx-10 fznLgY",
            },
            "children": Array [
              Object {
                "data": "iPhone 8",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "text",
              },
            ],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": [Circular],
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "alt": "iPhone 8",
            "class": "Products__ProductImage-sc-189ksbx-7 fsrlVZ",
            "src": "https://aerolab-challenge.now.sh/images/iPhone8-x1.png",
          },
          "children": Array [],
          "name": "img",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {
              "class": "Products__BuyLogo-sc-189ksbx-2 bvVWja",
              "src": "/icons/buy-blue.svg",
            },
            "children": Array [],
            "name": "img",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
              "src": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
              "src": undefined,
            },
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "alt": undefined,
            "class": undefined,
            "src": undefined,
          },
          "x-attribsPrefix": Object {
            "alt": undefined,
            "class": undefined,
            "src": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "Products__Category-sc-189ksbx-9 jLhkHD",
        },
        "children": Array [
          Object {
            "data": "Phones",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": Object {
          "attribs": Object {
            "class": "Products__ProductName-sc-189ksbx-10 fznLgY",
          },
          "children": Array [
            Object {
              "data": "iPhone 8",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": null,
          "parent": [Circular],
          "prev": [Circular],
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "Products__Break-sc-189ksbx-8 hPsqUm",
          },
          "children": Array [],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {
              "alt": "iPhone 8",
              "class": "Products__ProductImage-sc-189ksbx-7 fsrlVZ",
              "src": "https://aerolab-challenge.now.sh/images/iPhone8-x1.png",
            },
            "children": Array [],
            "name": "img",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "class": "Products__BuyLogo-sc-189ksbx-2 bvVWja",
                "src": "/icons/buy-blue.svg",
              },
              "children": Array [],
              "name": "img",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
                "src": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
                "src": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "alt": undefined,
              "class": undefined,
              "src": undefined,
            },
            "x-attribsPrefix": Object {
              "alt": undefined,
              "class": undefined,
              "src": undefined,
            },
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
      Object {
        "attribs": Object {
          "class": "Products__ProductName-sc-189ksbx-10 fznLgY",
        },
        "children": Array [
          Object {
            "data": "iPhone 8",
            "next": null,
            "parent": [Circular],
            "prev": null,
            "type": "text",
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": Object {
          "attribs": Object {
            "class": "Products__Category-sc-189ksbx-9 jLhkHD",
          },
          "children": Array [
            Object {
              "data": "Phones",
              "next": null,
              "parent": [Circular],
              "prev": null,
              "type": "text",
            },
          ],
          "name": "div",
          "namespace": "http://www.w3.org/1999/xhtml",
          "next": [Circular],
          "parent": [Circular],
          "prev": Object {
            "attribs": Object {
              "class": "Products__Break-sc-189ksbx-8 hPsqUm",
            },
            "children": Array [],
            "name": "div",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": [Circular],
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "alt": "iPhone 8",
                "class": "Products__ProductImage-sc-189ksbx-7 fsrlVZ",
                "src": "https://aerolab-challenge.now.sh/images/iPhone8-x1.png",
              },
              "children": Array [],
              "name": "img",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": Object {
                "attribs": Object {
                  "class": "Products__BuyLogo-sc-189ksbx-2 bvVWja",
                  "src": "/icons/buy-blue.svg",
                },
                "children": Array [],
                "name": "img",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": [Circular],
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                  "src": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                  "src": undefined,
                },
              },
              "type": "tag",
              "x-attribsNamespace": Object {
                "alt": undefined,
                "class": undefined,
                "src": undefined,
              },
              "x-attribsPrefix": Object {
                "alt": undefined,
                "class": undefined,
                "src": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
          "type": "tag",
          "x-attribsNamespace": Object {
            "class": undefined,
          },
          "x-attribsPrefix": Object {
            "class": undefined,
          },
        },
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {
      "class": undefined,
    },
    "x-attribsPrefix": Object {
      "class": undefined,
    },
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
